src/base/object_factory.hpp

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2008 Search Solution Corporation. All rights reserved by Search Solution.
+ *
+ *   This program is free software; you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation; either version 2 of the License, or
+ *   (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+/*
+ * object_factory.hpp
+ */
+
+#ifndef _OBJECT_FACTORY_HPP_
+#define _OBJECT_FACTORY_HPP_
+
+#ident "$Id$"
+
+#include <cassert>
+#include <functional>
+#include <map>
+
+/*
+ * factory of objects:
+ * main purpose is to be used with unpacking objects beloning to a class hierarchy:
+ * 'object_base' is the base class having derived classed 'object_type'
+ * each 'object_type' is identified by a value of 'object_key'
+ */
+namespace cubbase
+{
+  template <typename object_key, typename object_base> class factory
+    {
+      public:
+        template <typename object_type>
+        int register_creator (object_key k,
+                              std::function <object_base*()> object_creator = default_creator <object_type>)
+          {
+            typename std::map<object_key, std::function <object_base*()> >::iterator it = m_creators_map.find (k); 

```

Comment:
Why "typename" is used here? Try "auto", it will drastically reduce the verbosity of the declaration
2nd thought: do you really need it? Isn't  enough: `m_creators_map[k] = object_creator;` ?
- return type is almost useless (even you don't used it in the following code)
- assert(false) will prevent users of this class to change an existing association between key and creator but what if they need it? (it is a possible scenario for a factory); I think such logic should be outside this class, not inside

Reply:
When I wrote the unit test code, I found myself mistakenly copy-pasting the "register" calls with the same key. How to prevent this ?
I don't see the utility in changing the factory during runtime (new creator for same key); I think one creates it one time, if he needs other contents for factory, instantiates a new factory.

Reply:
In this case I would say to use the assert() during development only and remove it after you finish, otherwise it will be a restricted version of a factory
Or, If you really need this functionality for your case then you could give it a specific name, UnpackFactory for example, or use the return type to see if it was something already associated with a key...
Utility?... I don't have an example at hand but I know that generic & constraints don't go well together